balancer/pickfirst/pickfirstleaf/pickfirstleaf.go

```diff
@@ -179,14 +198,15 @@ func (b *pickfirstBalancer) resolverErrorLocked(err error) {
 	// The picker will not change since the balancer does not currently
 	// report an error. If the balancer hasn't received a single good resolver
 	// update yet, transition to TRANSIENT_FAILURE.
-	if b.state != connectivity.TransientFailure && b.addressList.size() > 0 {
+	if b.connectivityState != connectivity.TransientFailure && b.addressList.size() > 0 {
 		if b.logger.V(2) {
 			b.logger.Infof("Ignoring resolver error because balancer is using a previous good update.")
 		}
 		return
 	}
 
-	b.cc.UpdateState(balancer.State{
+	b.connectivityState = connectivity.TransientFailure
```

Comment:
> So, this change has nothing to do with adding the health listener? 

This isn't related to the health listener.

> This was just missed out previously?

Yes.

Should we have a test for this scenario?

After trying to write a test, I realized that the change due to this line is un-detectable from outside the balancer. Due to the following check just above this line,
https://github.com/grpc/grpc-go/blob/0775031253f395d1008f80da5f0582b8871f143e/balancer/pickfirst/pickfirstleaf/pickfirstleaf.go#L179-L187

if execution reaches here, either or both of the following will be true:
*  `b.state == connectivity.TransientFailure`: In this case, setting b.state back to TF is a no-op.
* `b.addressList.size() == 0`: Since the address list is empty, there is no way for the ClientConnState to transition other than a resolver update. The SubConn will be transitioned to CONNECTING on receiving a non-empty resolver update due to the `prevAddrCount == 0` line. 
https://github.com/grpc/grpc-go/blob/0775031253f395d1008f80da5f0582b8871f143e/balancer/pickfirst/pickfirstleaf/pickfirstleaf.go#L271-L279

Even though the line seems to be a no-op, I think it's good to keep it because it makes it easier to see that `b.connectivityState` is in sync with the state reported to the channel.

Reply:
Makes sense. Thanks for looking into it.